plugins/parsers/csv/parser_test.go

```diff
@@ -827,3 +827,222 @@ corrupted_line
+		MetadataSeparators: []string{":=", ",", ":", "="},
+		MetadataTrimSet:    " #'",
+	}
+	err = p.Init()
+	require.NoError(t, err)
+	require.Empty(t, p.metadataTags)
+	m = p.parseMetadataRow("# this is a not matching string")
+	require.Nil(t, m)
+	m = p.parseMetadataRow("# key1 : value1 \r\n")
+	require.Equal(t, m, map[string]string{"key1": "value1"})
+	m = p.parseMetadataRow("key2=1234\n")
+	require.Equal(t, m, map[string]string{"key2": "1234"})
+	m = p.parseMetadataRow(" file created : 2021-10-08T12:34:18+10:00 \r\n")
+	require.Equal(t, m, map[string]string{"file created": "2021-10-08T12:34:18+10:00"})
+	m = p.parseMetadataRow("file created: '2021-10-08T12:34:18'\r\n")
+	require.Equal(t, m, map[string]string{"file created": "2021-10-08T12:34:18"})
+}
+
+func TestParseCSVFileWithMetadata(t *testing.T) {
+	p := &Parser{
+		HeaderRowCount:     1,
+		SkipRows:           2,
+		MetadataRows:       4,
+		Comment:            "#",
+		TagColumns:         []string{"type"},
+		MetadataSeparators: []string{":", "="},
+		MetadataTrimSet:    " #",
+	}
+	err := p.Init()
+	require.NoError(t, err)
+	testCSV := `garbage nonsense that needs be skipped
+
+# version= 1.0
+
+    invalid meta data that can be ignored.
+file created: 2021-10-08T12:34:18+10:00
+timestamp,type,name,status
+2020-11-23T08:19:27+10:00,Reader,R002,1
+#2020-11-04T13:23:04+10:00,Reader,R031,0
+2020-11-04T13:29:47+10:00,Coordinator,C001,0`
+	expectedFields := []map[string]interface{}{
+		{
+			"name":      "R002",
+			"status":    int64(1),
+			"timestamp": "2020-11-23T08:19:27+10:00",
+		},
+		{
+			"name":      "C001",
+			"status":    int64(0),
+			"timestamp": "2020-11-04T13:29:47+10:00",
+		},
+	}
+	expectedTags := []map[string]string{
+		{
+			"file created": "2021-10-08T12:34:18+10:00",
+			"test":         "tag",
+			"type":         "Reader",
+			"version":      "1.0",
+		},
+		{
+			"file created": "2021-10-08T12:34:18+10:00",
+			"test":         "tag",
+			"type":         "Coordinator",
+			"version":      "1.0",
+		},
+	}
+	// Set default Tags
+	p.SetDefaultTags(map[string]string{"test": "tag"})
+	metrics, err := p.Parse([]byte(testCSV))
+	require.NoError(t, err)
+	for i, m := range metrics {
+		require.Equal(t, expectedFields[i], m.Fields())
+		require.Equal(t, expectedTags[i], m.Tags())
+	}
+
+	p = &Parser{
+		HeaderRowCount:     1,
+		SkipRows:           2,
+		MetadataRows:       4,
+		Comment:            "#",
+		TagColumns:         []string{"type", "version"},
+		MetadataSeparators: []string{":", "="},
+		MetadataTrimSet:    " #",
+	}
+	err = p.Init()
+	require.NoError(t, err)
+	testCSVRows := []string{
+		"garbage nonsense that needs be skipped",
+		"",
+		"# version= 1.0\r\n",
+		"",
+		"    invalid meta data that can be ignored.\r\n",
+		"file created: 2021-10-08T12:34:18+10:00",
+		"timestamp,type,name,status\n",
+		"2020-11-23T08:19:27+10:00,Reader,R002,1\r\n",
+		"#2020-11-04T13:23:04+10:00,Reader,R031,0\n",
+		"2020-11-04T13:29:47+10:00,Coordinator,C001,0",
+	}
```

Comment:
Why defining the same twice? Couldn't `p.Init()` be run a second time to reset internal counters? Also testCSVrows and the resulting tests could be reused imho.

Reply:
No, `Init()` will not reset the counters because we are using the variable set by the config (i.e. the `MetadataRows` field) directly.

Reply:
Okay, but still defining the test csv rows and the resulting tests twice seems unnecessary and error prone.

Reply:
@etycomputer could you change to define the test csv only once please

Reply:
Still not changed or given any comment on why not..

Reply:
@srebhan I discussed with @etycomputer about this on Slack. Do you agree we should only test if the processor can handle csv's with metadata in this test only and maybe add another test that checks the compatibility of different line endings?

Reply:
I'm fine either way.